Challenge resolution

index325 opened the pull request:
Olá, Jean,

Aqui está o PR.

Agradeço muito pela oportunidade e pelo desafio. Aprendi muitas coisas com ele.

Espero que o meu trabalho corresponda as tuas expectativas, dei o meu melhor.

Em caso de qualquer dúvida, pode me comunicar 😊

index325 committed:
Challenge resolution

jeandias commented on ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+
+  def upvote
+    @post.upvotes += 1
+    @post.save
```
```suggestion
    @post.update(upvotes: @post.upvotes + 1)
```
existe alguma razão em atribuir o valor e depois salvar o objeto, em vez de simplesmente fazer numa só operação?

index325 replied:
Eu estou acostumado a setar os valores no objeto antes e salvar por último. Mas a sugestão de fazer em uma só linha, me parece ótima. Vou aderir.

jeandias commented on ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def index
+    posts = Post.all.sort_by do |post|
```
não seria melhor um campo na BD que guarda esse valor calculado e assim remover essa responsabilidade e custo do endpoint?

index325 replied:
Posso criar um campo com o rating que é atualizado sempre quando o post recebe um upvote ou downvote e ordenar por ele. Vou fazer e já faço o commit

index325 replied:
Corrigido nos últimos commits

jeandias commented on ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  private
+
+  def post_params
+    params.require(:post).permit(:title, :upvotes, :downvotes)
```
qual a razão de criar o post já com votos?

index325 replied:
Para evitar que o post comece com votos nulos. Mas posso fazer uma mudança para que, se caso o post seja criado sem votos, ele inicialize com 0

jeandias commented on ix_challenge_list_posts_by_rating/app/models/post.rb:
```diff
@@ -0,0 +1,5 @@
+class Post < ApplicationRecord
+  validates :title, presence: true
+  validates :upvotes, presence: true
```
o que acontece se eu tentar criar um post sem votos?

index325 replied:
Neste caso, se você nâo enviar votos, o sistema retornará um erro dizendo que o campo não pode ser blank.

jeandias commented on ix_challenge_list_posts_by_rating/config/routes.rb:
```diff
@@ -0,0 +1,6 @@
+  post '/posts', to: 'posts#create'
+  put '/upvote/:id', to: 'posts#upvote'
+  put '/downvote/:id', to: 'posts#downvote'
+  get '/posts', to: 'posts#index'
```
conheces outra forma "mais elegante" para escrever as rotas REST?

index325 replied:
No curso de Rails eu aprendi a fazer com resources. Porém com o resources, ele cria todas as rotas necessárias para um CRUD. Posso pesquisar um jeito melhor

index325 replied:
Implementei uma nova forma nos últimos commits :)

jeandias commented on ix_challenge_list_posts_by_rating/test/controllers/posts_controller_test.rb:
```diff
@@ -0,0 +1,77 @@
+
+class PostsControllerTest < ActionDispatch::IntegrationTest
+  setup do
+    @post = Post.create(id: 667784312, title: "Post", upvotes: 1, downvotes: 1)
```
porque criar os objetos informando o id?

index325 replied:
Pois, se não for informado o ID, o teste de criação irá falhar. Pois, os posts que não têm o ID informado seguirão a ordem do autoincrement. E quando for ocorrer o teste da criação, ele dirá que o ID 1 já existe

jeandias replied:
os testes não podem guardar valores. deve haver uma limpeza na BD sempre que os testes correm e nunca ficarmos dependentes de valor fixos.

index325 replied:
Corrigido nos últimos commits

jeandias commented on ix_challenge_list_posts_by_rating/test/controllers/posts_controller_test.rb:
```diff
@@ -0,0 +1,77 @@
+
+    assert_equal @ordered_post_1.id, response_data[0]["id"]
+    assert_equal @ordered_post_2.id, response_data[1]["id"]
+  end
```
porque não implementar um teste com o que está no enunciado do desafio e que prova que o objetivo foi atingido?

index325 replied:
Posso mudar os valores para que fiquem exatamente iguais ao do enunciado. No meu teste eu apenas verifiquei com valores diferentes, porém validando se o 1 era o index 0 e o 2, o index 1

jeandias replied:
o testes que estou a falar é o que o 600/400, 60/40 e 6/4 ficam nessa ordem, apesar de todos terem o mesmo rating

index325 replied:
Corrigido no último commit. O peso utilizado foi a quantidade de interações.

index325 committed:
Update ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb

Adotando sugestão feita por Jean

Co-authored-by: Jean Peixoto Dias <jeanpeixotodias@gmail.com>

goncalopc96 commented on ix_challenge_list_posts_by_rating/docker-compose.yml:
```diff
@@ -0,0 +1,17 @@
+    image: postgres:12
+    environment:
+      POSTGRES_USER: invoicexpress
+      POSTGRES_PASSWORD: 123456
```
tens alguma outra forma de passar esta password sem estar exposta?

index325 replied:
Vou remover a password do docker-compose.yml. As variáveis deverão ser passadas pelo comando:
```
docker-compose up -e POSTGRESS_PASSWORD=SUA_SENHA
```

index325 replied:
Tem sim, eu estou fazendo um commit com as variáveis sensíveis protegidas

jeandias commented on ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def upvote
+    @post.upvotes += 1
```
com a abordagem da linha inferior essa linha deixa de fazer sentido

index325 replied:
Feito nos últimos commits

jeandias commented on ix_challenge_list_posts_by_rating/app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  def downvote
+    @post.downvotes += 1
+    @post.save
```
aqui aplica-se o mesmo que o método anterior

index325 replied:
Feito nos últimos commits

index325 committed:
[gabriel-resolution]

 - Changed the way that the routes are created
 - Changed the rating code to the model
 - Starting to fix the tests
 - Applying some suggestions

index325 committed:
[gabriel-resolution]

 - Putting parameters on docker-compose.yml

index325 committed:
[gabriel-resolution]

 - Fixing sorting code
